File: android/app/src/main/java/com/instachat/app/webview/PageHardening.kt
package com.instachat.app.webview

/**
 * JavaScript injected into every page Instagram creates.
 *
 * The native navigation guard (see DmNavigationPolicy) only sees real document
 * navigations. Instagram's web client is a React SPA, so most in-app clicks
 * (Home, Explore, Reels, a profile avatar, ...) are client-side route changes
 * that never trigger a navigation. This script closes that gap.
 */
object PageHardening {

    fun getScript(messageHandlerName: String = "InstaChatHost"): String = """
        (() => {
          'use strict';

          const INBOX = 'https://www.instagram.com/direct/inbox/';
          const ALLOWED = /^\/(direct|accounts|challenge|ajax|api|graphql|oauth)(\/|$)/i;
          const MESSAGE_HANDLER = '${messageHandlerName}';

          const pathAllowed = (p) => ALLOWED.test(p);

          const hideBlockedRailLinks = () => {
            try {
              document.querySelectorAll('nav').forEach((rail) => {
                rail.querySelectorAll('a[href]').forEach((a) => {
                  let u = null;
                  try { u = new URL(a.href, location.href); } catch (e) { return; }
                  if (u && u.origin === location.origin && !pathAllowed(u.pathname)) {
                    a.style.display = 'none';
                  }
                });
              });
              document.querySelectorAll('svg[aria-label="Instagram"]').forEach((svg) => {
                const a = svg.closest('a');
                if (a) a.style.display = 'none';
              });
            } catch (e) { /* ignore */ }
          };

          const guard = () => {
            try {
              if (!pathAllowed(location.pathname)) {
                location.replace(INBOX);
                return;
              }
              hideBlockedRailLinks();
            } catch (e) { /* ignore */ }
          };

          const wrap = (fn) => function () {
            const result = fn.apply(this, arguments);
            setTimeout(guard, 0);
            return result;
          };
          try {
            history.pushState = wrap(history.pushState);
            history.replaceState = wrap(history.replaceState);
            window.addEventListener('popstate', () => setTimeout(guard, 0));
          } catch (e) { /* ignore */ }

          setInterval(guard, 500);

          // ---- In-app toasts ----
          const TOAST_HOLD_MS = 6000;
          let overlay = null;
          const lastMirroredText = new WeakMap();

          const ensureOverlay = () => {
            if (overlay && overlay.isConnected) return overlay;
            overlay = document.createElement('div');
            overlay.id = 'ica-toast-overlay';
            overlay.setAttribute('aria-live', 'polite');
            overlay.style.cssText = [
              'position: fixed',
              'right: 16px',
              'bottom: 16px',
              'z-index: 2147483647',
              'display: flex',
              'flex-direction: column',
              'gap: 8px',
              'pointer-events: none',
              'max-width: 420px'
            ].join(';');
            try {
              (document.body || document.documentElement).appendChild(overlay);
            } catch (e) { return null; }
            return overlay;
          };

          const mirrorToast = (node) => {
            const host = ensureOverlay();
            if (!host) return;
            try {
              const card = document.createElement('div');
              card.textContent = (node.textContent || '').trim();
              if (!card.textContent) return;
              card.style.cssText = [
                'pointer-events: auto',
                'background: rgba(38, 38, 38, 0.95)',
                'color: #ffffff',
                'padding: 12px 16px',
                'border-radius: 10px',
                'font-size: 14px',
                'line-height: 1.4',
                'font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif',
                'box-shadow: 0 6px 24px rgba(0, 0, 0, 0.35)',
                'opacity: 1',
                'transition: opacity 0.4s ease'
              ].join(';');
              host.appendChild(card);

              setTimeout(() => {
                card.style.opacity = '0';
                setTimeout(() => card.remove(), 450);
              }, TOAST_HOLD_MS);
            } catch (e) { /* ignore */ }
          };

          const scanForToasts = () => {
            try {
              if (!document.body) return;
              document.querySelectorAll(
                '[role="alert"], [role="status"], [aria-live="assertive"], [aria-live="polite"]'
              ).forEach((node) => {
                if (node.id === 'ica-toast-overlay' || overlay?.contains(node)) return;
                if (node.querySelector('[role="alert"], [role="status"], [aria-live]')) return;
                const text = (node.textContent || '').trim();
                if (text.length > 200) return;
                if (!text) { lastMirroredText.set(node, ''); return; }
                const rect = node.getBoundingClientRect();
                if (rect.width < 40 || rect.height < 12) return;
                if (lastMirroredText.get(node) === text) return;
                lastMirroredText.set(node, text);
                mirrorToast(node);
              });
            } catch (e) { /* ignore */ }
          };

          try {
            const mo = new MutationObserver(() => scanForToasts());
            mo.observe(document.documentElement, { childList: true, subtree: true });
            setInterval(scanForToasts, 250);
            scanForToasts();
          } catch (e) { /* ignore */ }

          // ---- Notch cover ----
          let notchCover = null;
          const ensureNotchCover = () => {
            try {
              if (!document.body) return;
              if (!notchCover || !notchCover.isConnected) {
                notchCover = document.createElement('div');
                notchCover.id = 'ica-notch-cover';
                notchCover.style.cssText = [
                  'position: fixed',
                  'left: 0',
                  'top: 0',
                  'width: 100%',
                  'height: 48px',
                  'background: #ffffff',
                  'z-index: 2147483646',
                  'pointer-events: none'
                ].join(';');
                document.body.appendChild(notchCover);
              }
              const bodyBg = getComputedStyle(document.body).backgroundColor;
              const htmlBg = getComputedStyle(document.documentElement).backgroundColor;
              const bg = bodyBg && bodyBg !== 'rgba(0, 0, 0, 0)' ? bodyBg : htmlBg;
              if (bg && bg !== 'rgba(0, 0, 0, 0)') notchCover.style.background = bg;
            } catch (e) { /* ignore */ }
          };
          setInterval(ensureNotchCover, 500);
          ensureNotchCover();

          // ---- Create button hiding ----
          const CREATE_ICON = /^(create|new post)$/i;
          const hideCreateButton = () => {
            try {
              document.querySelectorAll('svg[aria-label]').forEach((svg) => {
                if (!CREATE_ICON.test((svg.getAttribute('aria-label') || '').trim())) return;
                const item = svg.closest('a')
                  || svg.closest('div[role="button"]')
                  || svg.closest('button')
                  || svg.closest('li');
                if (item) item.style.display = 'none';
              });
              document.querySelectorAll('a, button, div[role="button"]').forEach((el) => {
                const label = el.getAttribute('aria-label') || el.getAttribute('title') || '';
                if (CREATE_ICON.test(label.trim())) el.style.display = 'none';
              });
            } catch (e) { /* ignore */ }
          };
          setInterval(hideCreateButton, 500);
          hideCreateButton();

          // ---- Title relay ----
          const sendTitle = () => {
            try {
              const title = document.title || '';
              if (window[MESSAGE_HANDLER] && typeof window[MESSAGE_HANDLER].postMessage === 'function') {
                window[MESSAGE_HANDLER].postMessage(JSON.stringify({ type: 'title', title: title }));
              }
            } catch (e) { /* ignore */ }
          };
          try {
            const titleEl = document.querySelector('title');
            if (titleEl) {
              new MutationObserver(sendTitle).observe(titleEl, { subtree: true, childList: true, characterData: true });
            }
          } catch (e) { /* ignore */ }
          setInterval(sendTitle, 1000);
          sendTitle();

          // ---- Force light mode (prevent Instagram dark theme) ----
          const forceLightMode = () => {
            try {
              let existingStyle = document.getElementById('ica-light-mode');
              if (existingStyle) existingStyle.remove();

              const style = document.createElement('style');
              style.id = 'ica-light-mode';
              style.textContent = [
                'html, body {',
                '  color-scheme: light !important;',
                '  background-color: #ffffff !important;',
                '}',
                ':root { --ig-primary-background: #ffffff; --ig-secondary-background: #fafafa; }',
                'textarea, input[type="text"], [contenteditable="true"],',
                '[role="textbox"], [aria-label="Message"],',
                '[aria-label="Message…"], [placeholder*="Message"] {',
                '  background-color: #fafafa !important;',
                '  color: #262626 !important;',
                '}',
                'div[style*="background-color"] {',
                '  /* handled below */',
                '}'
              ].join('\\n');
              (document.head || document.documentElement).appendChild(style);

              // Force colorScheme on documentElement
              document.documentElement.style.colorScheme = 'light';

              // Walk inline-styled elements and neutralise dark backgrounds
              document.querySelectorAll('div, section, main, aside, nav, header, footer').forEach((el) => {
                const bg = el.style.backgroundColor;
                if (!bg) return;
                // If the inline background is very dark, replace with white
                const m = bg.match(/\\d+/g);
                if (m && m.length >= 3) {
                  const r = parseInt(m[0]), g = parseInt(m[1]), b = parseInt(m[2]);
                  if (r < 60 && g < 60 && b < 80) {
                    el.style.backgroundColor = '#ffffff';
                  }
                }
              });
            } catch (e) { /* ignore */ }
          };
          setInterval(forceLightMode, 1000);
          forceLightMode();

          guard();
        })();
    """.trimIndent()
}

File: src/GramChat/Services/PageHardening.cs
namespace GramChat.Services;

/// <summary>
/// JavaScript injected into every page Instagram creates.
/// </summary>
/// <remarks>
/// The native navigation guard (see <see cref="NavigationPolicy"/>) only sees real document
/// navigations. Instagram's web client is a React single-page app, so most in-app clicks (a
/// Reels icon, an in-feed reel, a profile's "Reels" tab) are client-side route changes that
/// never trigger a navigation. This script closes that gap: it watches the History API and
/// force-returns the SPA to the Home feed whenever it lands on a Reels or Explore route, and it
/// hides every link/tab that points at one.
///
/// It is deliberately event-driven. The chat-only build ran four separate <c>setInterval</c>
/// loops (500 ms x2, 250 ms, 1000 ms); this version drives all of the DOM work from a single
/// <c>MutationObserver</c> collapsed into one animation frame, with one slow 2 s backstop for
/// route changes the observer cannot see. While a voice/video call is on screen it stops the
/// cosmetic DOM work entirely so the renderer's main thread stays free for WebRTC, and it
/// notifies the host so the shell can keep the screen awake during the call.
/// </remarks>
public static class PageHardening
{
    public const string Script = """
        (() => {
          'use strict';

          const HOME = 'https://www.instagram.com/';
          // Reels, Explore and every reel-adjacent route. Matched per path
          // segment so it also catches a profile's /username/reels/ tab and
          // /reels/audio/... .
          const BLOCKED = /(^|\/)(reel|reels|reelsaudio|clips|explore)(\/|$)/i;
          const TAB_LABEL = /^(reels?|clips|explore)$/i;
          const TOAST_SELECTOR = '[role="alert"], [role="status"], [aria-live="assertive"], [aria-live="polite"]';
          const CALL_LABEL = /^(end call|leave call|join call|turn (off|on) video|switch to (video|audio)|mute audio|unmute audio|decline call|accept call)$/i;
          const TOAST_HOLD_MS = 6000;

          const isBlocked = (p) => BLOCKED.test(p || '');

          // ---- tiny scheduler: collapse bursts of DOM work into one frame ----
          let scheduled = false;
          const schedule = (fn) => {
            if (scheduled) return;
            scheduled = true;
            const run = () => { scheduled = false; try { fn(); } catch (e) { /* ignore */ } };
            if (window.requestAnimationFrame) window.requestAnimationFrame(run);
            else window.setTimeout(run, 16);
          };

          // ---- host bridge: one place that knows how to talk to the shell ----
          const post = (payload) => {
            try { window.chrome?.webview?.postMessage(payload); } catch (e) { /* ignore */ }
          };

          // ================= routing guard =================
          const guard = () => {
            if (isBlocked(location.pathname)) {
              location.replace(HOME);
              return true;
            }
            return false;
          };

          const wrap = (fn) => function () {
            const result = fn.apply(this, arguments);
            setTimeout(guard, 0);
            return result;
          };
          try {
            history.pushState = wrap(history.pushState);
            history.replaceState = wrap(history.replaceState);
            window.addEventListener('popstate', () => setTimeout(guard, 0));
            window.addEventListener('hashchange', () => setTimeout(guard, 0));
          } catch (e) { /* ignore */ }

          // ================= page tidy =================
          // Hide every link/tab that points at Reels or Explore: the rail icons,
          // an in-feed reel card, a profile's "Reels" tab. Unrelated navigation
          // is left alone (the old chat-only build hid all of it).
          const tidy = () => {
            try {
              document.querySelectorAll('a[href]').forEach((a) => {
                let u = null;
                try { u = new URL(a.href, location.href); } catch (e) { return; }
                if (u.origin !== location.origin || !isBlocked(u.pathname)) return;
                const item = a.closest('li')
                  || a.closest('article')
                  || a.closest('[role="tab"]')
                  || a;
                item.style.display = 'none';
              });
              document.querySelectorAll('[role="tab"], [role="menuitem"]').forEach((el) => {
                if (TAB_LABEL.test((el.textContent || '').trim())) el.style.display = 'none';
              });
            } catch (e) { /* ignore */ }
          };

          // ================= call mode =================
          // While a call is on screen we stop the cosmetic DOM work entirely so
          // the renderer's main thread stays free for WebRTC, and tell the shell
          // so it can keep the screen awake / hold audio focus during the call.
          let callMode = false;
          let lastCallCheck = 0;
          const postCallState = (active) => post({ type: 'call', active: active });
          const detectCall = (force) => {
            const now = Date.now();
            if (!force && now - lastCallCheck < 750) return;
            lastCallCheck = now;
            let active = false;
            try {
              active = Array.from(document.querySelectorAll('[aria-label]')).some((el) => {
                const label = (el.getAttribute('aria-label') || '').trim();
                if (!CALL_LABEL.test(label)) return false;
                const r = el.getBoundingClientRect();
                return r.width > 0 && r.height > 0;
              });
            } catch (e) { active = false; }
            if (active !== callMode) {
              callMode = active;
              postCallState(active);
            }
          };

          // ================= in-app toasts =================
          // Instagram's own toast banners auto-dismiss almost immediately; mirror
          // every one into a fixed overlay and hold it for a few seconds so it
          // stays readable. OS-level (WebView2/WebKit) notifications are
          // unaffected - the shell already auto-grants the Notifications
          // permission.
          let overlay = null;
          const lastMirroredText = new WeakMap();

          const ensureOverlay = () => {
            if (overlay && overlay.isConnected) return overlay;
            overlay = document.createElement('div');
            overlay.id = 'gcm-toast-overlay';
            overlay.setAttribute('aria-live', 'polite');
            overlay.style.cssText = [
              'position: fixed', 'right: 16px', 'bottom: 16px', 'z-index: 2147483647',
              'display: flex', 'flex-direction: column', 'gap: 8px',
              'pointer-events: none', 'max-width: 420px'
            ].join(';');
            try { (document.body || document.documentElement).appendChild(overlay); } catch (e) { return null; }
            return overlay;
          };

          const mirrorToast = (node) => {
            const host = ensureOverlay();
            if (!host) return;
            try {
              const card = document.createElement('div');
              card.textContent = (node.textContent || '').trim();
              if (!card.textContent) return;
              card.style.cssText = [
                'pointer-events: auto', 'background: rgba(38, 38, 38, 0.95)', 'color: #ffffff',
                'padding: 12px 16px', 'border-radius: 10px', 'font-size: 14px', 'line-height: 1.4',
                'font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif',
                'box-shadow: 0 6px 24px rgba(0, 0, 0, 0.35)', 'opacity: 1', 'transition: opacity 0.4s ease'
              ].join(';');
              host.appendChild(card);
              setTimeout(() => {
                card.style.opacity = '0';
                setTimeout(() => card.remove(), 450);
              }, TOAST_HOLD_MS);
            } catch (e) { /* ignore */ }
          };

          const scanForToasts = () => {
            try {
              if (!document.body) return;
              document.querySelectorAll(TOAST_SELECTOR).forEach((node) => {
                if (node.id === 'gcm-toast-overlay' || overlay?.contains(node)) return;
                // Containers that hold a live region are skipped - the inner node
                // gets mirrored instead (avoids double toasts).
                if (node.querySelector(TOAST_SELECTOR)) return;
                const text = (node.textContent || '').trim();
                if (text.length > 200) return;
                if (!text) { lastMirroredText.set(node, ''); return; }
                const rect = node.getBoundingClientRect();
                if (rect.width < 40 || rect.height < 12) return;   // screen-reader only
                if (lastMirroredText.get(node) === text) return;
                lastMirroredText.set(node, text);
                mirrorToast(node);
              });
            } catch (e) { /* ignore */ }
          };

          // ================= title relay =================
          // Relay Instagram's document title (it carries the unread count) to the
          // shell, but only when it actually changes.
          let lastTitle = null;
          const sendTitle = () => {
            const title = document.title || '';
            if (title === lastTitle) return;
            lastTitle = title;
            post({ type: 'title', title: title });
          };

          // ================= one observer drives all of it =================
          let mo = null;
          const onDomChanged = () => {
            detectCall(false);
            if (callMode) return;   // keep the main thread free during calls
            tidy();
            scanForToasts();
          };
          const startObserver = () => {
            if (mo || !document.documentElement) return;
            try {
              mo = new MutationObserver(() => schedule(onDomChanged));
              mo.observe(document.documentElement, { childList: true, subtree: true, characterData: true });
            } catch (e) { /* ignore */ }
          };
          try {
            const titleEl = document.querySelector('title');
            if (titleEl) {
              new MutationObserver(sendTitle).observe(titleEl, { subtree: true, childList: true, characterData: true });
            }
          } catch (e) { /* ignore */ }

          // Slow backstop: correct a route change we never observed, re-attach
          // the observer after a full reload, and refresh the title. A single
          // timer replaces the four separate setInterval loops of the old build.
          setInterval(() => {
            guard();
            startObserver();
            detectCall(false);
            if (!callMode) { tidy(); scanForToasts(); }
            sendTitle();
          }, 2000);

          // Initial pass.
          guard();
          startObserver();
          detectCall(true);
          if (!callMode) { tidy(); scanForToasts(); }
          sendTitle();
        })();
        """;
}
